pyBluzelle/communication.py

```diff
@@ -0,0 +1,140 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import websocket
+import json
+import logging
+import base64
+import sys
+import random
+
+try:
+    from pyBluzelle.proto import bluzelle_pb2
+    from pyBluzelle.proto import database_pb2
+except ImportError as e:
+    raise ImportError("{}\n\nTo generate Bluzelle protobuf modules:\n"
+                      "\n"
+                      "$ protoc --python_out=./pyBluzelle/proto --proto_path=proto/proto bluzelle.proto database.proto audit.proto\n".format(e.message))
+
+logger = logging.getLogger(__name__)
+
+
+class Connection(object):
+
+    def __init__(self, host, port, uuid):
+        """
+        Create a connection to a Bluzelle database
+        :param host: hostname for the Bluzelle node
+        :param port: port of the Bluzelle node
+        :param uuid: uuid of the database to connect
+        """
+        self._host = host
+        self._port = port
+        self._uuid = uuid
+
+    def _send_request(self, host, port, msg):
+        ws = websocket.create_connection("ws://{}:{}".format(host, port))
+        msg.db.header.db_uuid = self._uuid
+        msg.db.header.transaction_id = random.randint(1, sys.maxint)
+
+        req = {}
+        req["bzn-api"] = "database"
+        req["msg"] = base64.b64encode(msg.SerializeToString())
+
+        logger.debug("Sending: {}".format(msg))
+        logger.debug("-" * 60 + '\n')
+        var = json.dumps(req)
+        ws.send(json.dumps(req))
+        resp = database_pb2.database_response()
+        resp.ParseFromString(ws.recv())
+
+        if resp.WhichOneof('success') == 'redirect':
+            host = resp.redirect.leader_host
+            port = resp.redirect.leader_port
+            logger.debug("redirecting to leader at {}:{}...\n".format(host, port))
+            resp = self._send_request(host, port, msg)
+        else:
+            logger.debug("Response: \n{}".format(resp))
+            logger.debug("-" * 60 + '\n')
+
+        ws.close()
+
+        return resp
+
+    def create(self, key, value):
```

Comment:
There are a piece of code that keeps poling the result after the create operation is done because Bluzelle did not guarantee read after write consistency (If you create a thing, you can't immediately read it). Seems like that piece of code has been removed. Does Bluzelle guarantee read after write consistency on the server side now? or we just leave it to the application layer to deal with inconsistency?

Reply:
Sounds good